ea-factories/src/factories/balance.ts

```diff
@@ -0,0 +1,116 @@
+import {
+  ExecuteFactory,
+  Config,
+  DataResponse,
+  Account,
+  SequenceResponseData,
+} from '@chainlink/types'
+import objectPath from 'object-path'
+import { Validator, AdapterError } from '@chainlink/external-adapter'
+import { util } from '@chainlink/ea-bootstrap'
+
+const DEFAULT_DATA_PATH = 'addresses'
+const DEFAULT_CONFIRMATIONS = 6
+
+const WARNING_UNSUPPORTED_PARAMS = 'No Operation: this provider does not support'
+const ERROR_MISSING_ADDRESS = 'No Operation: address param is missing'
+
+export const addresses = (accounts: Account[]) =>
+  accounts.filter((a) => !a.warning).map((a) => a.address)
+
+export type IsSupported = (coin: string, chain: string) => boolean
+export type BalancesResponse = DataResponse<Account[], any>
+export type GetBalance = (account: Account, config: BalanceConfig) => Promise<BalancesResponse>
+export type GetBalances = (accounts: Account[], config: BalanceConfig) => Promise<BalancesResponse>
+
+export type BalanceConfig = Config & {
+  confirmations?: number
+  shouldOverwrite?: boolean
+  verbose?: boolean
+  isSupported: IsSupported
+  getBalance?: GetBalance
+  getBalances?: GetBalances
+}
+
+const requireArray = (jobRunID: string, dataPath: string, data: any) => {
+  const inputData = <Account[]>objectPath.get(data, dataPath)
+
+  // Check if input data is valid
+  if (!inputData || !Array.isArray(inputData) || inputData.length === 0)
+    throw new AdapterError({
+      jobRunID,
+      message: `Input, at '${dataPath}' path, must be a non-empty array.`,
+      statusCode: 400,
+    })
+
+  return inputData
+}
+
+const toValidAccount = (jobRunID: string, account: Account, config: BalanceConfig): Account => {
+  // Is it possible to process?
+  if (!account.address)
+    throw new AdapterError({
+      jobRunID,
+      message: ERROR_MISSING_ADDRESS,
+      statusCode: 400,
+    })
+
+  // Defaults
+  if (!account.chain) account.chain = 'mainnet'
+  if (!account.coin) account.coin = 'btc'
+
+  // Should we process?
+  if (!config.shouldOverwrite && typeof account.balance === 'number') return account
+
+  // Do we support processing?
+  const supported = config.isSupported(account.coin, account.chain)
+  if (!supported)
+    return { ...account, warning: WARNING_UNSUPPORTED_PARAMS + ` ${account.chain} ${account.coin}` }
+
+  // If warning, clear and continue to processing
+  const { warning, ...accNoWarning } = account
+  return accNoWarning
+}
+
+const toGetBalances = (getBalance?: GetBalance) => (accounts: Account[], config: BalanceConfig) => {
+  if (!getBalance) throw new Error('Get Balance function not supplied')
+  return accounts.map(async (acc) => {
+    if (acc.warning) return { result: [acc] }
```

Comment:
This is too late to filter warnings. If we do it here, we don't do it in the `config.getBalances` method.

There is also the problem of not keeping the original order for the sequence, grouping is messing up the order, and we need to restore it in the end.

First, we should filter out all not supported accounts before grouping:

```javascript
const supportedAccounts =  accounts.filter(a => !a.warning)
const groups = Array.from(util.groupBy<string, Account>(supportedAccounts, key).values())
const requests = ...
const responses = ...
```

Now that we have the  results back, we need to build an index of balances:

```javascript
const balances = new Map(responses.flatMap((r: any) => r.result).map(a => [a.address, a.balance])))
```

Then we build the response, using the original order:

```javascript
const data: SequenceResponseData<Account> = {
  responses: responses.map((r: any) => r.payload),
  result: accounts.map(a => {...a, balance: a.warning ? a.balance || balances.get(a.address)})
}
```

Reply:
This is a good suggestion, thanks. I had the warnings getting filtered _within_ `config.getBalances`, but that would be better abstracted away. Plus maintaining the order seems better too.